### ISSUE TITLE
[LaunchDarkly] identify error: LaunchDarklyTimeoutError: identify timed out after 5 seconds on IOS

### ISSUE THREAD

**Issue by ryan-brs** [launchdarkly/js-core#516]
**Describe the bug**
[LaunchDarkly] identify error: LaunchDarklyTimeoutError: identify timed out after 5 seconds

**To reproduce**
Use this component as a Provider
```
import {
  LDProvider,
  ReactNativeLDClient,
  AutoEnvAttributes,
} from "@launchdarkly/react-native-client-sdk";
import { useEffect } from "react";

const LaunchDarklyProvider = ({ children }: PropsWithChildren): ReactNode => {

  const client = new ReactNativeLDClient(MOBILE_KEY, AutoEnvAttributes.Disabled);

  useEffect(() => {
    const setup = async (): Promise<void> => {
      if (rowKey && token) {
        await client
          .identify({
            kind: 'multi',
            practice: {
              kind: 'practice',
              key: rowKey,
            },
          })
          .then(() => console.log('identify finished'))
          .catch((e: unknown) => console.log('client error', e));
      }
    };

    setup();
  }, [rowKey, token]);

  return <LDProvider client={client}>{children}</LDProvider>;
};

export default LaunchDarklyProvider;
```

**Expected behavior**
The error should not show and should be able to get the correct variation value.

**Logs**
LOG  error: [LaunchDarkly] identify error: LaunchDarklyTimeoutError: identify timed out after 5 seconds

**Also enabled debug mode and get these logs**
 debug: [LaunchDarkly] [EventSource] Will open new connection in 0 ms.
 LOG  info: [LaunchDarkly] Closed LaunchDarkly stream connection

**SDK version**
"@launchdarkly/react-native-client-sdk": "^10.2.1"

**Language version, developer tools**
"react-native": "0.74.3"
"expo": "51.0.18"
"expo-router": "3.5.17"

**OS/platform**
IOS simulator 17.5
Real device 17.5.1

**Additional context**
Can see the context from the website context section

**Comment by kinyoklion** [comment]
Hello @ryan-brs,

The start and identify methods support a configurable timeout that you can provide. The goal of the timeout is to ensure applications do not wait indefinitely for initialization. If an applications waits indefinitely for initialization, then it can an take indefinite time for the user to be able to use an application.

You can add a timeout to the identify options.
```
, { timeout: 10 }
```

The SDK continues to attempt to initialize in the background regardless of the timeout. If your application doesn't block anything important waiting for the LD SDK, then you could set it to a large number to effectively never timeout.

It does tend to be the case that iOS seems to have more timeouts. We don't have any application specific code for iOS which would impact this. 

The timeout also uses a custom type, so if you do not care about it timing out you could discard that type of exception.

I will look into the defaults to see if they need adjusted.

Thank you,
Ryan

**Comment by ryan-brs** [comment]
> Hello @ryan-brs,
> 
> The start and identify methods support a configurable timeout that you can provide. The goal of the timeout is to ensure applications do not wait indefinitely for initialization. If an applications waits indefinitely for initialization, then it can an take indefinite time for the user to be able to use an application.
> 
> You can add a timeout to the identify options.
> 
> ```
> , { timeout: 10 }
> ```
> 
> The SDK continues to attempt to initialize in the background regardless of the timeout. If your application doesn't block anything important waiting for the LD SDK, then you could set it to a large number to effectively never timeout.
> 
> It does tend to be the case that iOS seems to have more timeouts. We don't have any application specific code for iOS which would impact this.
> 
> The timeout also uses a custom type, so if you do not care about it timing out you could discard that type of exception.
> 
> I will look into the defaults to see if they need adjusted.
> 
> Thank you, Ryan
> 
> _Filed internally as 250330_

Hey Ryan,

Thanks for you reply, I have tried different timeout(even 300) but no lucky still get the same error

**Comment by kinyoklion** [comment]
If you have a user who opens your app, and there is not already a cached context, then it has to connect to LaunchDarkly to initialize. This presents many conditions. Lets say I open your app and I have no cell service, 5 minutes would still be within the time I may not have cell service.

So you will always get timeouts within a human timescale. I personally wouldn't consider it an error or log it as an error. It is like any other failed request, aside from the fact that we will automatically continue retrying for you.

There are several things that would be a problem:
- If the SDK doesn't eventually initialize, but internet is available
- If you get the timeout after 5 seconds, even though you set the timeout to 5 minutes.
- If the context is the same as a context that previously connected and initialized, then the result should already be cached. So it should not timeout.
- It is saying that it timed out after 5 seconds, even though it took 5 minutes.

**Comment by ryan-brs** [comment]
> If you have a user who opens your app, and there is not already a cached context, then it has to connect to LaunchDarkly to initialize. This presents many conditions. Lets say I open your app and I have no cell service, 5 minutes would still be within the time I may not have cell service.
> 
> So you will always get timeouts within a human timescale. I personally wouldn't consider it an error or log it as an error. It is like any other failed request, aside from the fact that we will automatically continue retrying for you.
> 
> There are several things that would be a problem:
> 
> * If the SDK doesn't eventually initialize, but internet is available
> * If you get the timeout after 5 seconds, even though you set the timeout to 5 minutes.
> * If the context is the same as a context that previously connected and initialized, then the result should already be cached. So it should not timeout.
> * It is saying that it timed out after 5 seconds, even though it took 5 minutes.

Pretty sure the connection is good as everything works well if I switch back to sdk 8.0.1. And context get sent to the contexts dashboard.

The timeout log looks fine as it is saying timeout error after the time I set.

I can also see the log "Identifying {kind: multi", rowKey: "rowkey"} so I'm assuming the initialize has been done?
but this log is followed by two other logs
 LOG  debug: [LaunchDarkly] [EventSource] Will open new connection in 0 ms.
 LOG  info: [LaunchDarkly] Closed LaunchDarkly stream connection
And timeout error will show after

**Comment by kinyoklion** [comment]
@ryan-brs 

Thank you for the detail. We will take a look.

Thank you,
Ryan

_Filed internally as 250413_

**Comment by ayushmahajan12** [comment]
Hi @ryan-brs 

Can you please let us know what is the impact on user's related to the timeout issue.
Over important release on hold due to the timeout failure.

<img width="864" alt="Screenshot 2024-07-18 at 5 53 01 PM" src="https://github.com/user-attachments/assets/2193e60d-6617-447c-99b1-07e5d71ada22">

**Comment by ramakula** [comment]
I'm sorry but this is a huge problem with the new JS SDK. We are switching back to native SDK. It is not possible to identify multiple times with the new SDK, it times out whereas the old native SDK never used to timeout. This error cannot be ignored because the real user identify fails and users are stuck with anonymous user flags.

<img width="1106" alt="Screenshot 2024-08-20 at 16 08 30" src="https://github.com/user-attachments/assets/06eab502-6581-474b-a7d7-0bce34fa7946">

It is so easy to reproduce by sending different context data 4 to 5 times:

<img width="1694" alt="Screenshot 2024-08-20 at 10 33 54" src="https://github.com/user-attachments/assets/55269f6b-46b3-4874-adfc-0ea3ca07bff8">


Everything works perfectly with the old native SDK version 9.3.0. We will not migrate to the new SDK unless this is fixed.

**Comment by kinyoklion** [comment]
@ramakula Have you tried setting the timeout longer. The old SDK did not timeout because it was not a capability that it had. If you do not care about the timeout, then you can catch it. It doesn't affect the operation of the SDK.

This SDK does timeout, but all the timeout does is cause the function to reject the promise. The identify completes in the background, but this prevents the possibility of indefinitely waiting in the case that LaunchDarkly cannot be reach.

We generally recommend less than 15 seconds, but if you are experiencing long initialization times, then you can go longer than that.

I will again check that longer timeouts are working. We will also consider increasing the default.

Thank you,
Ryan

**Comment by kinyoklion** [comment]
There is another consideration which is if you identify several contexts without waiting between, then the timeout applies to them individually. So if you identify 5 contexts and there isn't network access, then you get 5 timeouts.

I did verify the timeout does respect the time and the associated message will show the correct time.

**Comment by ramakula** [comment]
@kinyoklion I've tried up to 15 seconds. Timeout itself is not a big deal but we are seeing reports from the users that their features are not available and we found out that they are set to anonymous user. I'll try 30 seconds and get back to you with complete LD logs and why identify is set to anonymous etc. I'll also reach out to our main point of contact to setup a call so I can share more details with you guys. Appreciate all the tips.

**Comment by krizpoon** [comment]
In my case, there is a library in our project that intercepts traffic using the XMLHttpRequest object. Once the library is disabled, LD works again. That may explain why the native library works but not the RN library.

**Comment by kinyoklion** [comment]
Hello @krizpoon,

Is the library internal to your project, or part of a publicly available library? (Aside from flipper and expo issues which are already covered in the README.)

If it is a public library, then we would like to know which one, so we can add additional content to the readme, and potentially get the issue addressed.

Thank you,
Ryan

**Comment by krizpoon** [comment]
Hey @kinyoklion, it's a public library: https://www.npmjs.com/package/msw

**Comment by kinyoklion** [comment]
@krizpoon Thank you. It doesn't look like msw can maintain the streaming capability of the original request. It looks like it is converting it to a fetch and then emulating the XmlHttpRequest behavior once it reads a full response. (I didn't exhaustively look through it, but this is my impression from a cursor look.)

I am not sure how feasible it would be to change it to use a readable stream from the fetch, which could support streaming. The alternative used in expo is to check if the request expects a streamed response and then to not intercept it.

This is generally the problem most of these interception libraries (including that in flipper and expo make), is that HTTP responses need to be read in completion. When real HTTP responses can be read indefinitely as a stream. So they connect to our streaming endpoint and attempt to read the "entire" body, but that will not complete (at least until there is an error or a timeout).

**Comment by krizpoon** [comment]
Thanks @kinyoklion. Another problem we are facing is that the network logging feature on BrowserStack is blocking the SSE traffic as well. Is there any known solution or work around please?

**Comment by krizpoon** [comment]
Thanks @kinyoklion! Just realised that `polling` mode is now supported in the latest SDK release.

**Comment by lucasmendesdeliver** [comment]
Also having the same issue, any solution on that?

**Comment by CrystalCodes01** [comment]
I’m also encountering the LaunchDarklyTimeoutError: identify timed out after 5 seconds issue in my React Native app with Expo, using @launchdarkly/react-native-client-sdk@10.9.3.

My Context:

- The issue occurs when a user’s network disconnects and reconnects, triggering the identify process again.
- I’ve implemented a network listener (@react-native-community/netinfo) to detect reconnections and reinitialize the LaunchDarkly client. Despite this, the timeout still happens, and users are sometimes assigned an anonymous context instead of their actual identity.

Observations & Error Logs:

- [LaunchDarkly] Closed LaunchDarkly stream connection
- [LaunchDarkly] identify error: LaunchDarklyTimeoutError: identify timed out after 5 seconds
- Cached contexts don’t seem to mitigate the issue consistently during reconnections.

Questions:

- Are there updates or workarounds being explored to address this issue, especially regarding reconnection scenarios?
- Would reverting to an earlier SDK version (e.g., 9.x) resolve these timeouts, as suggested by others?
- Are there additional debugging steps or best practices to improve reliability in disconnected and reconnected states?

Thank you for any guidance. Happy to provide further logs or context if needed!

**Comment by kinyoklion** [comment]
Hello @CrystalCodes01,

The identify will timeout when there isn't network. It still will continue to retry and eventually connect. The timeout is just to prevent waiting indefinitely on a promise. If you are handling the exception, then the only effect should be logs. The timeout is just for the promise returned from identify, not the underlying operation. 

> I’ve implemented a network listener (@react-native-community/netinfo) to detect reconnections and reinitialize the LaunchDarkly client. 

This will likely just cause more logs as the SDK will already be attempting to connect using an exponential backoff strategy. What does "reinitialize" mean here? If it just means "identify", then it is just duplicate effort, but if it means a new client instance, then tha could be causing issues if the previous instance is not specifically closed. 

> and users are sometimes assigned an anonymous context instead of their actual identity.

I am very interested in this because it sounds like a bug. There isn't anything related to identify, other than identifying an anonymous context, that should ever result in any anonymous context. What is the observed behavior here?

> Cached contexts don’t seem to mitigate the issue consistently during reconnections.

If you identify a context and it was previously cached, then the SDK loads those cached values and starts connecting in the background. A context is identified by its kinds and their associated keys. If you have a multi-context and any of the keys change, then that is a different context instance from the identify perspective and will not be able to load any cached values.

We will eventually be adding network info handling in order to defer even attempting to connect. But the timeout would not be addressed, as mentioned it is just the callsite. So if you make a call to identify, and we have not finished it in 5 seconds, then it will still raise the timeout. 

If you are not intending to wait for the promise, then you could set the timeout effectively indefinitely. The timeout is accepted as an option to the identify call.

Thank you,
Ryan

**Comment by kinyoklion** [comment]
In regards to reproduction, while a timeout is 100% reproducible, because it is the expected behavior of the API. 

But I've never been able to reproduce any situation where it doesn't recover once a network is available (as this operation is independent of the callsite timeout), or where it doesn't load a previously cached context (when there is a matching context already cached).

**Comment by lucasmendesdeliver** [comment]
@ryan-brs I found the solution for this problem, in my case in using expo SDK 51, and I was calling the identify more than one time.

The solution was just make sure the identify method is called only once.

**Comment by kinyoklion** [comment]
An extra note on cached contexts. That feature will only currently work if `@react-native-async-storage/async-storage` is available or if passing in custom storage in the configuration.

**Comment by brianangulo** [comment]
When is this expected to be fixed? It seems there is actually an issue here but is not clear what it is. We have an issue where LD doesn't initialize when using a couple of libraries that intercept network requests for SSL Pinning(@bam-tech) and logging (Quantum Metric). From our perspective it just appears to time out and not initialize which holds our Splash screen up way too long. Any thoughts or suggestions?

**Comment by kinyoklion** [comment]
It isn't ever expected that the identify method could not timeout. That said a timeout doesn't generally represent a terminal condition in normal usage, effectively it represents the maximum time you want to present that splash screen, for example. You can additionally control it to suit your needs.

When a timeout happens the SDK continues in the background.

If this happens every initialization then that is clearly a problem. I cannot say specifically what bam-tech and quantum metrics do to the HTTP request. It is a super common situation for interception libraries to incorrectly handle SSE. The SDK, by default, uses a streaming connection to initialize and receive updates from LaunchDarkly. Often interception libraries expect HTTP requests to have a finite duration, and will not respond to the call with incremental data sent over the stream. This will present the SDK from initializing. 

You can test this easily by setting the initial connection mode of the SDK to polling. If this works, then it is likely there is a defect (or intentional decision) in one of those libraries.

```
const featureClient = new ReactNativeLDClient(MOBILE_KEY, AutoEnvAttributes.Enabled, {
  debug: true,
  applicationInfo: {
    id: 'ld-rn-test-app',
    version: '0.0.1',
  },
  initialConnectionMode: 'polling'
});
```

We previously experienced this problem with expo (https://github.com/expo/expo/issues/27526), and they have resolved it, and before that flipper.

As with the expo issue the easiest way for a library to avoid this problem is to check for `text/event-stream`. Though it is better to not assume HTTP requests aren't streaming.

Thank you,
Ryan

**Comment by brianangulo** [comment]
> It isn't ever expected that the identify method could not timeout. That said a timeout doesn't generally represent a terminal condition in normal usage, effectively it represents the maximum time you want to present that splash screen, for example. You can additionally control it to suit your needs.
> 
> When a timeout happens the SDK continues in the background.
> 
> If this happens every initialization then that is clearly a problem. I cannot say specifically what bam-tech and quantum metrics do to the HTTP request. It is a super common situation for interception libraries to incorrectly handle SSE. The SDK, by default, uses a streaming connection to initialize and receive updates from LaunchDarkly. Often interception libraries expect HTTP requests to have a finite duration, and will not respond to the call with incremental data sent over the stream. This will present the SDK from initializing.
> 
> You can test this easily by setting the initial connection mode of the SDK to polling. If this works, then it is likely there is a defect (or intentional decision) in one of those libraries.
> 
> ```
> const featureClient = new ReactNativeLDClient(MOBILE_KEY, AutoEnvAttributes.Enabled, {
>   debug: true,
>   applicationInfo: {
>     id: 'ld-rn-test-app',
>     version: '0.0.1',
>   },
>   initialConnectionMode: 'polling'
> });
> ```
> 
> We previously experienced this problem with expo ([expo/expo#27526](https://github.com/expo/expo/issues/27526)), and they have resolved it, and before that flipper.
> 
> As with the expo issue the easiest way for a library to avoid this problem is to check for `text/event-stream`. Though it is better to not assume HTTP requests aren't streaming.
> 
> Thank you, Ryan

Thank you @kinyoklion! Setting the mode to polling does solve the issue we were experiencing. Do you think we can rely on setting connection mode to polling as a permanent solution or should we investigate the root cause further?

**Comment by kinyoklion** [comment]
@brianangulo Polling will work, but it is less responsive to changes than using streaming. It polls on an interval, so you receive flag changes at most on that interval. Versus streaming where we push changes to the client.

It will work, but it would be good to know which of the instrumentation libraries is causing the issue, and we may be able to report it with them to get is resolved for other users of SSE (and other streaming functionality).

Thank you,
Ryan

**Comment by Bardiamist** [comment]
I want just fetch all flags on app start and close connection.

But we noticed that it often doesn't work.

Here is my implementation:

```
const launchDarkly = new ReactNativeLDClient(config.launchDarklyKey, AutoEnvAttributes.Disabled);

launchDarkly.identify({
  kind: 'user',
  key: userId,
  }, {
    waitForNetworkResults: true,
  });

const allFlags = launchDarkly.allFlags();

launchDarkly.close();
```

I added logs where sorted keys alphabetically:

```
if (allFlags != null) {
  const sortedAllFlags = {};

  Object.entries(allFlags).sort((
    [firstKey],
    [secondKey],
  ) => firstKey.localeCompare(secondKey)).forEach(([key, value]) => {
    sortedAllFlags[key] = value;
  });

  const sortedAllFlagsString = JSON.stringify(sortedAllFlags, null, 2);
}
```

`sortedAllFlagsString` is `{}` it means that `allFlags()` returned empty
